### PR TITLE
[SYSTEMML-540] Added a rewrite to support a common tensor operation (sum over channels)

### PR DESCRIPTION
**Pull request by niketanpansare**
- Added a rewrite to convert `out = rowSums(matrix(colSums(A), rows=C, cols=HW))` to `out = channel_sums(A)` when `nrow(A) > 1` and exectype is CP or GPU.
- This avoids unnecessary intermediates, GPU-CP-GPU transfer (for reshape) and additional kernel invocation cost.
- When we move to a higher CuDNN version, we can replace the custom `channel_sums` kernel with possibly more optimized CuDNN reduce tensor kernel.
- Added the corresponding CPU and GPU tests.
- Updated `T_MAX(val)` to `MAX()`. Interestingly enough, `nvcc` was smart enough to remove the parameter automatically, hence the ptx remained the same after the change :)

@nakul02 can you please review this PR ?

### CODE REVIEW

**Comment by niketanpansare** [comment]
Let's mark this as WIP until we switch to CuDNN 7 and use more optimized `gpu_channel_sums`.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2214/

**Comment by mboehm7** [comment]
wrt to CP, it would be good to use a kahan plus for consistency with sum; also the benefits in CP are probably limited because reshape dense by row is just a meta data operation.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2216/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2218/

**Comment by niketanpansare** [comment]
As initial checkpoint, this PR eliminates unnecessary `reshape` in sentence CNN for [channel_sum](https://github.com/apache/systemml/blob/master/scripts/nn/util.dml#L42) and hence `D2H+H2D` cost:
Before this PR:
```
  6  rshape [42:16-42:55]                         129.243  141588                                                                                                       aqrd[75.808s,141588], aqmd[0.008s,46400], aqms[0.044s,95188], rlsi[0.105s,141588], rlswr[6.172s,0]
  7  gpu_uark+ [42:2-42:56]                        71.966  141588                                                             aqrd[0.034s,46400], aqrs[0.078s,95188], rrowk[8.792s,141588], s2d[4.605s,95188], sync[0.618s,141588], ad[0.920s,141588], H2D[23.032s,141588]
  8  rand [50:4-50:68]                             60.176   11799                                                                                                                                                                      aqmd[0.008s,11799], rlswr[0.018s,0]
  9  gpu_uack+ [42:16-42:55]                       41.416  139200                                                                                                                                            rcolk[38.500s,139200], sync[0.632s,139200], ad[1.629s,139200]
 10  gpu_-* [61:9-61:9]                            39.205  235980                                              aqrs[0.010s,11593], H2D[1.301s,11600], sync[23.549s,235980], daxpy[2.859s,235980], D2D[3.588s,235980], aqrd[0.000s,7], s2d[0.566s,11593], ad[2.198s,235980]
```

With this PR:
```
  6  rand [50:4-50:68]                             60.057   11799                                                                                                                                                                      aqmd[0.014s,11799], rlswr[0.015s,0]
  7  gpu_channel_sums [42:2-42:56]                 50.757  139200                                                                    rcolk[38.242s,139200], a[0.002s,6], r[0.045s,139194], sync[0.415s,139200], rrowk[9.405s,139200], az[0.888s,139200], ad[0.956s,139200]
  8  gpu_-* [61:9-61:9]                            38.728  235980                                             sync[24.591s,235980], aqrs[0.011s,11590], H2D[1.314s,11600], daxpy[2.502s,235980], D2D[3.238s,235980], aqrd[0.000s,10], s2d[0.608s,11590], ad[1.868s,235980]
```

@nakul02 I am ready to merge this PR if you don't have any comments.

File: src/main/cpp/kernels/SystemML.cu [hunk 1]
```diff
@@ -20,7 +20,7 @@
 /**********************************
 When updating a kernel or adding a new one,
 please compile the ptx file and commit it:
-nvcc -ptx -arch=sm_30 SystemML.cu
+nvcc -ptx -arch=sm_30 --std c++11 SystemML.cu
```
Comment (nakul02) [inline src/main/cpp/kernels/SystemML.cu]: There is makefile in this directory. Please update that with the C++11 flag too.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/ConvolutionCPInstruction.java [hunk 2]
```diff
@@ -297,6 +320,65 @@ public void processBiasMultiplyInstruction(ExecutionContext ec) throws DMLRuntim
 		ec.setMatrixOutput(getOutputVariableName(), outputBlock, getExtendedOpcode());
 	}
 	
+	public void processChannelSumsInstruction(ExecutionContext ec) throws DMLRuntimeException {
+		MatrixBlock input = ec.getMatrixInput(input1.getName(), getExtendedOpcode());
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/runtime/instructions/cp/ConvolutionCPInstruction.java]: Are you also going to add a multithreaded version for this later?

File: src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCUDA.java [hunk 3]
```diff
@@ -322,6 +322,37 @@ public static void reluBackward(GPUContext gCtx, String instName, MatrixObject i
...
+		int N = toInt(input.getNumRows());
+		int cols = toInt(input.getNumColumns());
+		if(cols != C*HW) {
+			throw new DMLRuntimeException("Incorrect parameters, number of columns " + cols + " != " + C + "*" + HW);
+		}
+		Pointer imagePointer = getDensePointer(gCtx, input, instName);
+		Pointer outputPointer = getDensePointer(gCtx, outputBlock, instName);
+		
+		// We can replace this with CuDNN tensor reduce
+		Pointer tmp = gCtx.allocate(instName, cols*sizeOfDataType);
```
Comment (nakul02) [inline src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCUDA.java]: Is this temporary space accounted for in your memory estimation ?

Comment (niketanpansare) [reply src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixCUDA.java]: Done in latest commit.

File: src/main/java/org/apache/sysml/runtime/instructions/cp/ConvolutionCPInstruction.java (hunk 2 above)
Comment (niketanpansare) [reply src/main/java/org/apache/sysml/runtime/instructions/cp/ConvolutionCPInstruction.java]: No, it is purely a memory bound operation.

File: src/main/cpp/kernels/SystemML.cu (hunk 1 above)
Comment (niketanpansare) [reply src/main/cpp/kernels/SystemML.cu]: Done.

**Comment by niketanpansare** [comment]
Any additional comments ?

**Comment by nakul02** [comment]
Nothing more to add for the GPU side of things. 
GPU related part LGTM 👍

**Comment by niketanpansare** [comment]
Thanks @nakul02 ... I will merge :+1:

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2222/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2222/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2222/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.parfor/ParForDataPartitionLeftIndexingTest/testParForDataPartitionLeftindexingRowSparseMR/'><strong>org.apache.sysml.test.integration.functions.parfor.ParForDataPartitionLeftIndexingTest.testParForDataPartitionLeftindexingRowSparseMR</strong></a></li></ul>

**Pull request closed without merging**